File: Data Structures & Algorithms/minimum-remove-to-make-valid-parentheses/submission-1.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        oc = cc = 0

        for c in s:
            cc += c == ")"
        
        res = []
        for c in s:
            if c == "(":
                if oc == cc:
                    continue
                oc += 1
            elif c == ")":
                cc -= 1
                if oc == 0:
                    continue
                oc -= 1
            res.append(c)
        return "".join(res)

File: Data Structures & Algorithms/k-closest-points-to-origin/submission-1.java
class Solution {
    public class Coords {
        int x;
        int y;
        int distanceSquared;  // Store squared distance to avoid sqrt computation

        public Coords(int x, int y) {
            this.x = x;
            this.y = y;
            this.distanceSquared = x * x + y * y;
        }
    }

    public class CoordsComparator implements Comparator<Coords> {
        public int compare(Coords p1, Coords p2) {
            return Integer.compare(p1.distanceSquared, p2.distanceSquared);
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Coords> pq = new PriorityQueue<>(new CoordsComparator());
        int[][] res = new int[k][2];
        for(int[] point : points) {
            pq.add(new Coords(point[0], point[1]));
        } 

        for(int i = 0; i < k; i++) {
            Coords c = pq.poll();
            res[i] = new int[] {c.x, c.y};
        }

        return res;
    }
}
